sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -139,6 +139,10 @@ class ReturnHelper {
   size_t *param_value_size_ret;
 };
 
+// save discovered pi_devices & pi_platforms for quick return
+static std::vector<pi_device> piDevicesCache;
```

Comment:
Please, avoid adding global objects with c'tors and d'tors where possible. https://llvm.org/docs/CodingStandards.html#do-not-use-static-constructors

Reply:
@romanovvlad , Do you have any suggestion how to avoid the global objects for this cache implementaion?
I need to use this object across calls to piDevicesGet() and piDevicesRelease().

Reply:
No, I have no any simple and reliable solution. I can only suggest having std::array with some "max size" here. It less likely will cause crashes. Why device and platform caching is needed, isn't it a bug in the users code if expensive resource is acquired several times instead of reusing?

Reply:
> Why device and platform caching is needed, isn't it a bug in the users code if expensive resource is acquired several times instead of reusing?
It is not a bug in the user code. 
Note that pi_device is just a wrapper of low-level ZE_DEVICE_HANDLE in L0 RT.
As I mentioned in the main description, the current implementation of piDevicesGet() and piPlatformsGet() are implemented to always return new pi_devices and pi_platforms every time these PI APIs are called.
It causes a problem on the SYCL RT side because from device_impl and platform_impl have no way to distinguish whether they are actually the same sycl::device or not, and same sycl::platform or not.
This PR fixes this issue by returning always the same pi_devices and pi_platforms for the same ZE_DEVICE_HANDLE and ZE_PLATFORM_HANDLE.

Reply:
> No, I have no any simple and reliable solution. I can only suggest having std::array with some "max size" here. It less likely will cause crashes.
Are you saying std::array is not a class and does not have constructors and destructors?

Reply:
@romanovvlad the idea of SYCL is to simplify programming, not to duplicate assembly-level programming. We have already C and assembler for this. :-)

Reply:
@bso-intel to avoid a global complex constructor you can for example have a `std::unique_ptr` that you initialize from a static variable initialization in a function using it for the first time or something like that.
Actually something like line 455 sounds good too.

Reply:
@keryell Just curious, who do you think is guilty that the following code is slow, developer or standard C++ library?:

```
    1 #include <fstream>
    2
    3 int main() {
    4   constexpr int N = 1200000;
    5   //std::fstream A("/tmp/file");
    6   for(int I = 0; I < N; ++I) {
    7     std::fstream A("/tmp/file");
    8     A << I << std::endl;
    9   }
   10   return 0;
   11 }
```

Reply:
> to avoid a global complex constructor you can for example have a std::unique_ptr that you initialize from a static variable initialization

If we use unique_ptr we will still have problem with lifetime at it's not defined when the object is destroyed with regard to the application.

Reply:
> @keryell Just curious, who do you think is guilty that the following code is slow, developer or standard C++ library?:

This is a good point. :-)

Reply:
> If we use unique_ptr we will still have problem with lifetime at it's not defined when the object is destroyed with regard to the application.

I see, your are not caching an object used behind some SYCL object, just some PI objects outside of any SYCL connection... Then, my question is: why this caching is required?

This is related to some Khronos internal discussion https://gitlab.khronos.org/sycl/Specification/-/issues/306 about observability.

Reply:
> Then, my question is: why this caching is required?

According to the description, it should 1) help with determining whether sycl::device is the same and 2) help to avoid calling expensive L0 RT.

@bso-intel Could you please elaborate more on how the patch should help with 1)? Objects of the sycl::device are compared by, well, comparing pointers to device_impl rather than pi handles they encapsulate.

Reply:
I tried to explain how this patch solves the sycl::device equality issue.
The issue of sycl::device equality is caused by the fact that L0 plugin always creates a new pi_device object for the same ZE_DEVICE_HANDLE. 
So, every time SYCL RT calls piDevicesGet(), L0 plugin always returns a vector of different pi_device objects.
If L0 plugin always returns the vector of same pi_devices, sycl::device can easily distinguish the returned pi_devices are the same as other sycl::device.
@jbrodman may answer this question better. Please help me here if I missed something. :)

Reply:
If you add caching in SYCL layer, `piDevicesGet()` will be called only once and so there is no need for caching at this level.
@jbrodman I am interested by your feedback on the hot issue 306 in Khronos SYCL.